### PR TITLE
#167727632 Create email and push Notifications for new travel requests

### PR DESCRIPTION
**Pull request by chrismeeky**
#### What does this PR do?
Create email and push Notifications for new travel requests
#### Description of Task to be completed?

- When a new travel request is created by a direct report, send email and push notification (using socket.io API) to his line manager to allow him make approval decisions. 

- create an endpoint to render the frontend to test the push notification feature

- create a new table to save new messages

#### How should this be manually tested?
* Clone the repo and `cd` into the project directory
* Run `yarn install` to install the project's dependency
* Run `yarn migrate` to create new tables
* Run `yarn seed` to seed to the database
* Run `yarn start` to start the app
* Open your browser and visit `http://localhost:[PORT]/api/v1/managerNotification/<:ManagerId>` . This will render the front end for test purposes
* Open **Postman** and make a post request to `localhost:[PORT}/api/v1/request/book_trip` or any route for booking a trip.
* Book the trip with a user whose line manager corresponds with that entered in the browser.
* A push notification is sent from your browser as soon as the request is made as illustrated in the screenshot below. The Message count is incremented in realtime on each push and saved in the database

**Notification on FireFox**

Screenshot from 2019-09-08 11-01-00

**Notification on Chrome**
Screenshot from 2019-09-08 15-30-22

* An email notification is sent to the line manager to allow him make an approval decision. Below is a screenshot containing the sample email

Screenshot from 2019-09-08 23-32-13

#### What are the relevant pivotal tracker stories?
[#167727632](https://www.pivotaltracker.com/story/show/167727636)

#### Any background context you want to add?

- Since web sockets are not supported by postman, you will need to test the feature using any browser
- To test the email feature, you need to replace the line manager's email in the seed file with yours

- Only managers to the direct report can receive new request notifications

- Manager must be subscribed to email notification in order to receive one

### CODE REVIEW

**Review by Daymorelah — CHANGES_REQUESTED** [review]
The `notifications.js` file that you created in the `Utils` folder is causing ESLint errors which means you didn't place the file where it's supposed to be. looking at your implementation, the file is meant to be in the controller's folder not in the utils folder. you cannot be importing models in files that are in the utils folder.

File: src/controllers/NotificationController.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+   * Test real-tim in-app notification
+   * Route: POST: api/v1/
+   * @param {object} req - HTTP Request object
+   * @param {object} res - HTTP Response object
+   * @return {res} res - HTTP Response object
+   * @memberof Notification
+   */
+  static async notify(req, res) {
+    const { id } = req.params;
+    const { Message } = models;
```
Comment (Daymorelah) [inline src/controllers/NotificationController.js]: Why are you doing getting the models here?. This is not how we have been importing models in the codebase.

File: src/utils/notifications.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+          managerName: firstName,
+          requester: user.firstName,
+          requestId: dataValues.id,
+          origin: dataValues.origin,
+          destination: dataValues.destination,
+          flightDate: dataValues.flightDate,
+          reason: dataValues.reason,
+          type
+        });
+      }
```
Comment (Daymorelah) [inline src/utils/notifications.js]: how do we know am email was actually sent here?. Please ensure that an email was being sent before going o to the next logic

File: src/utils/notifications.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+      await Message.create({
+        message,
+        userId: id,
+        lineManager: user.dataValues.lineManager,
+        type: 'creation'
+      });
```
Comment (Daymorelah) [inline src/utils/notifications.js]: How can we know for certain that a message was being created in the db?

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/controllers/NotificationController.js (hunk 1 above)
Comment (chrismeeky) [reply src/controllers/NotificationController.js]: That was a mistake. Thank you

File: src/utils/notifications.js (hunk 2 above)
Comment (chrismeeky) [reply src/utils/notifications.js]: Okay

File: src/utils/notifications.js (hunk 3 above)
Comment (chrismeeky) [reply src/utils/notifications.js]: I will re-factor

**Pull request closed without merging**